### ISSUE TITLE
betweenness centrality results

### ISSUE THREAD

**Issue by jexp** [neo4j-contrib/neo4j-graph-algorithms#282]
from[ neo4j-users slack](https://neo4j-users.slack.com/archives/C5C4JRFK7/p1502898665000543): 

Hi, I'm trying to run the Betweenness Centrality algorithm. I'm
<img width="851" alt="screen shot 2017-08-16 at 8 45 22 am" src="https://user-images.githubusercontent.com/67427/29392385-721f8884-82f5-11e7-91a2-f7c0e7eee63a.png">
 using the graph shown in the screenshot: match (a:Node)-[b:Edge]->(c:Node) return *.  To run the algorithm, I'm executing the following: CALL algo.betweenness("Node", "Edge", {write:true, stats:true, writeProperty:'centrality'})
YIELD nodes, minCentrality, maxCentrality, sumCentrality. The problem is that the results don't seem correct to me. When there are multiple equivalent nodes between 2 other nodes (e.g. 2,3,4 between 1&5 and 6,7 between 5 &8 ), only one of the equivalent nodes is getting a centrality value greater than 0. My understanding is that they should all have the same centrality.

**Comment by tomasonjo** [comment]
I have tested this scenario. I think the user was visualizing IDs and not betweenness weights as they are nicely ordered from 1 to 8.

Create data:
```

CREATE (a:Node{id:"A"}),
       (b:Node{id:"B"}),
       (c:Node{id:"C"}),
       (d:Node{id:"D"}),
       (e:Node{id:"E"}),
       (f:Node{id:"F"}),
       (g:Node{id:"G"}),
       (h:Node{id:"H"})
CREATE (a)-[:EDGE]->(b),
       (a)-[:EDGE]->(c),
       (a)-[:EDGE]->(d),
       (b)-[:EDGE]->(e),
       (c)-[:EDGE]->(e),
       (d)-[:EDGE]->(e),
       (e)-[:EDGE]->(f),
       (e)-[:EDGE]->(g),
       (f)-[:EDGE]->(h),
       (g)-[:EDGE]->(h)
```

Run the algo: 

```
CALL algo.betweenness("Node", "EDGE", {write:true, stats:true, writeProperty:'centrality'})
YIELD nodes, minCentrality, maxCentrality, sumCentrality
```

Results:


![betweenness1](https://user-images.githubusercontent.com/19948365/29409171-5ae971b0-834b-11e7-8268-5d247b99a737.png)

Now this results are also strange on the first look. I think that when there are more possible shortest path algo just picks one consistently, so running algo many times does not produce different results.


Run the algo.exp1()

```
CALL algo.betweenness.exp1("Node", "EDGE", {write:true, stats:true, writeProperty:'centra'})
YIELD nodes, minCentrality, maxCentrality, sumCentrality
```

Results:

![betweeness2](https://user-images.githubusercontent.com/19948365/29409619-65b63d06-834d-11e7-8003-828df4b504a5.png)


This results look very nice, have no problem with many shortest paths. Maybe this should be the default.

**Comment by tomasonjo** [comment]
I have also tested apoc.betweenness() for

> if concurrency parameter is set (and >1) ParallelBetweennessCentrality is used

With concurrency params 1 and 4 and they both yield same results.

**Comment by tomasonjo** [comment]
Tested it in gephi to see what does it return. We get the same results as `.exp1()` version. So there is a bug with the original betwenness algo. Should I make a separate issue ?

![gephi](https://user-images.githubusercontent.com/19948365/29412434-179bcde6-8359-11e7-9b0c-b702d3071e4d.png)

**Comment by adstruble** [comment]
Hi @tomasonjo,
I'm the user that reported the issue. Sorry if my original post wasn't clear, you were correct that I was visualizing IDs, my screenshot was just intended to show the graph contents. The results that I explained in the text portion of my post were the same incorrect results that you reproduced. I'm a little confused about your posts though, what would the separate issue be? The only issue that I'm seeing here is that Betweenness centrality using predecessor sets has incorrect results.

**Comment by tomasonjo** [comment]
@adstruble 

You are right, betweenness centrality using predecessor sets is the only issue here. I thought it was a random user reporting this and it wasn't clear yet that predecessor sets wasn't working right. There is no need for separate issue.

**Comment by jexp** [comment]
@tomasonjo perhaps a separate issue makes sense?

**Comment by jexp** [comment]
@mknblch ^^^

**Comment by mknblch** [comment]
Thanks for the info. 
I debugged the original Brandes implementation and found an error in the calculation of sigma (shortest path counts). This is most likely the cause of this bug. The same implementation exists in ParallelBrandes so this one will fail as well. Im working on it..

**Comment by mknblch** [comment]
I made a testcase for this problem in `org.neo4j.graphalgo.algo.BetweennessCentralityIntegrationTest_282`. Should be fixed with #316.
